### DISCUSSION TITLE
Reworking the Organs tab

### DISCUSSION THREAD

**Discussion by oleg68** [GrandOrgue/grandorgue#955] [General]
There are several issues related to working with a list of the organs and organ packages.

- #786
- #886
- #950

After implementing #948 there will be one tab ``Organs``.

The basic idea is to have the single list of registered organs with additional features.

I want to get the end users' opinions what information they want to see in the common list and what actions they want to have in this window.

Especially I ask @paolo2504 and @larspalo to share their ideas here.

**Comment by paolo2504** [comment]
Oleg, i reviewed #948 past days and i agree with almost everything. HOwever, i believe these changes, maybe not difficult to apply, are many so i'll review quietly the all stuff to be able to give you a reasonable anwser.
At first, however, i wish to summarize the first couple of things that arise in my mind,

1. to be able to identify the cache files and the organs they belong to
2. automatic deletion of cache/cbm files when organs are deleted/unregistered from Settings (no matter they are package organs or not)
3. to be able to install organ packages no matter compression used and extension (rar/zip/gzip); otherwise, i guess that the function "install package" is useless. I mean: most samplesets available on the net (P.Grabowsky,countrebombarde,odfgrandorgue and so on) are compressed with normal or high ratio and come with .zip or .rar extension (.exe on odfgrandorgue French site, in this case you have to execute to unpack and create the folder structure. Note: i have still not found how to directly unpack these files...).
However, GO only supports 0-ratio compression (no compression) for files with .orgue extension. This limits, imho, the real advantage to use the function "install organ package".
Note: i believe that this functionality, if fixed and optimized, would be of great help as opposite to "File/open". If you fix this function, then you can have File/Load, Install package (or even toggle it because GO does authomatically recognize .orgue files even under the Samplesets folder set in Options tab).
4. Option File/open recent , to be honest, was never used by me and could be removed

**Reply by larspalo** [reply]
@paolo2504, @oleg68

1. Handling of the files must be better from within GO. There's no need to "be able to identify what organ" a certain file in the file system belongs to besides from within GO (Organs tab). All handling should be done from within GO with all possible options (remove, copy/backup, etc). Invalidated cache files should be possible to track down and delete. As should any possibly orphaned cache/settings files.
2. Maybe it would be prudent to allow a choice to delete the files and not just do it without a warning, but yes it should absolutely be possible to manage all files handled by GO from within GO, in my opinion.
3. @paolo2504 your text has very little to do with the Organs tab per se. It just seems you've still not understood the new GO package format, nor what the respective menu option is for. The first paragraph in the wiki page here should put you on the right track: https://github.com/GrandOrgue/grandorgue/wiki/Organ-package-creation-(GrandOrgueTool) and see https://github.com/GrandOrgue/grandorgue/wiki#quickstart-guide for why both options are there...
4. The fact that you've never used the feature doesn't mean it necessarily should be removed. But if we unify all organs (old and new style), then they perhaps should be handled with the same priority and in the same way.

**Reply by paolo2504** [reply]
Thanks Lars. Some thought. 

1/2. Fully agree with you but consider that remove (delete) currently does not delete cache files, combinations and organ settings. Honestly I think there is no reason to leave corresponding orphaned files when users do decide to delete organs ! Just clean them as they become useless.. Regarding copy/backup , I don’t see those features at present. Maybe you’re speaking of future ones you’re thinking to develope ? In that case, I don’t understand then real advantage : why copy or backup an organ ? Sorry I don’t understand. 
3. Guess I have not been able to explain well: my apologies. What I mean with organ packages is that I do like this feature but I don’t understand when to use it. First, I can’t use it because grandorguetool doesn’t work under windows (I opened a ticket for that). Second, the majority of the samplesets abailable on the net comes in rar/zip format with normal/high compression ratio. GO is not able to read them. It looks like the “install organ package” feature is addressed to users that want to create their own organs but this is not my case. I would enjoy GOnbeing able to directly read for example Friesach sampleset just downloaded from P. Grabowsky’s website but this not possible today: I have to unpack it (renaming the .rar file in .orgue doesn’t work because GO requires a 0-compression ratio). At present , unless being a organ package creatore (and I am not…) the only way to open for example Friesach is unzip it and do File/open: for this reason I wrote that install organ package feature is useless at present: in fact I have 70-80 samplesets more or less all rar/zip ones and only two (Pitea and the Demo) with .orgue format. Hope I have been able to explain my pov. 
4. Agree with you. Oleg asked to comment that was I did. For me file/ open recent is not necessary but I fully understand that someone else would think of it be useful. No problem.

**Reply by larspalo** [reply]
1,2. is my general thoughts of how I think it should be done, not so much how it currently works.
3. Read the links I provided again carefully. Almoast no new format "organ packages" are available, sadly enough. (I have provided PiteaMHS and BureaChoir) Most other "packages" of GO samplesets only qualify as "old style" samplesets and no, they cannot be installed, nor do they qualify as "organ packages" in GO terms. In time I'll either re-work some of my older samplesets or re-record them completely and distribute them as new style GO organ packages.

**Reply by vpoguru** [reply]
The GO packages format is fantastic! Gernot Wurst (PROSPECTUM) recently published the Zoeblitz sample set as GO organ package.

**Reply by jason-barba** [reply]
( @larspalo I have re-worked my extension of your Kalvträsk extension to an organ package if you are interested. )

**Reply by larspalo** [reply]
@jason-barba Thanks, but I'll intend to also re-work the samples for that. Ideally, I'd like to completely re-record Kalvträsk organ, as I really enjoy that small gem and think it's worth the best of the best...

**Reply by jason-barba** [reply]
Great, looking forward to that. It is the one organ/sample set I use on my VPO :-)

**Comment by paolo2504** [comment]
Lars, to demonstrate how much I appreciate the “organ package” feature I say I’ll rework my samplesets as soon as the grandorguetool for Windows will be working. 
At present it does not. 
I have also tried winrar and 7zip (using 0-ratio compression that means store only) but GO is getting error on parsing content (“no end record found”).
As an alternative, I would greatly appreciate (in addition to have a working GrandOrgueTool) let GO read and manage directly the rar/zip files that we all find on internet. 
Thanks.

**Reply by larspalo** [reply]
It's perfectly possible to create .orgue packages (new style sampleset distribution) completely manually, so you don't absolutely have to wait for GrandOrgueTool to be fixed on windows. But in your case, it might be best just to wait until GrandOrgueTool is working under windows as it makes things much easier. As I'm only using Linux myself it's rather pointless to give you the exact commands and procedure to do this... But someone else might actually be able to use it (or we make an addition to the wiki) so here it goes:

This is how to create an .orgue package manually for a sampleset that's complete and ready for distribution. Nothing should be expected to be done to this package after it's created - if updating really is needed, a new package should be created and distributed - either as a new version or just as a replacement of the old.

1. First make absolutely sure that the paths look like `Pipe001=Subbas16\036-C.wav` and this includes all file paths, not only the samples.
2. Wavpack all the samples. In the root directory of the sampleset you can run this shell script: 
~~~
for a in `find -name \*.wav`; do wavpack -h -x6 $a -o $a.new && mv $a.new $a; done
~~~
3. Create an organindex.ini file that must be in the root directory of the sampleset. See GO help for full description, here's just an arbitrary example:
~~~
[General]
Title=Bureå Funeral Chapel sampleset version 20211225
OrganCount=2

[Organ001]
Filename=burea_gravkapell.organ
ChurchName=Bureå Gravkapell (Original)
OrganBuilder=Johannes Menzel
RecordingDetails=Recorded 2009 by Lars Palo

[Organ002]
Filename=burea_gravkapell_ext.organ
ChurchName=Bureå Gravkapell (Extended)
OrganBuilder=Johannes Menzel
RecordingDetails=Recorded 2009 by Lars Palo
~~~
4. Pack the whole sampleset root directory into a zip archive. For instance one can (in the root directory) issue the following command:
`zip -r ../BureaFuneralChapel . -0`
5. Rename the created archive with an .orgue extension and your done! (note that in my example above the archive iself will be created outside the root directory of the sampleset)

P.S. Old style sampleset distibution **must** be manually extracted as GO cannot read whatever kind of compression were used when the archives were created.

**Reply by vpoguru** [reply]
We should really try to promote the .orgue format as it creates very small installation packages and is very straightforward to use.

An important step could be to make the GrandOrgueTool a bit more comfortable.

**Reply by paolo2504** [reply]
I created a Windows batch script to automatize wavpacking. It works, i'll post it after some final testing.

There is an issue however with the packing (i used 7zip and winrar): GO seems not able to parse the organindex file.
I made a test with PiteaMHS20160115.orgue got from Lars's website.
I unpacked it then i repacked it again with NO modification at all using winrar and 7z with 0-compression.
GO is not able to parse it (see error below). Why ? organindex,ini is the same used by Lars and i suppose all .wav files were already wavpacked...  i suspect that zip tool under Linux works different than in Windows; perhaps some Linux expert can help.

btw: Lars-Oleg is it possible to post here the script ? Maybe with .txt extension so GitHub doesn’t block it (it’s a .cmd). 



![immagine](https://user-images.githubusercontent.com/50132661/148927675-5e2f10b2-bc69-4d78-9866-5e0097f6f825.png)

**Reply by jason-barba** [reply]
As I understand (please correct me if I am wrong), an .orgue file is a zip archive with 0-compression.  So therefor I don't understand "using winrar and 7z". Is 7z archives equal to a zip archive? or do you mean 7zip also in the quote? What is the purpose of winrar? You say above that you use the program 7zip; my 7zip opens .orgue files directly; I don't need winrar.

Could these ignorant questions have bearing on your problem?

**Reply by paolo2504** [reply]
My apologies Jason this morning I was in a hurry because I had to go to school (I’m a teacher) so I wrote post too much quickly…

Definitely .orgue files are zip archive with 0-compression (it’s called “store compression”). 
Normally I use winrar, sometimes 7z. They are similar. 
I wrote “winrar and 7z” but didn’t mean together! I was trying to understand why GO can’t parse orgue  files created with winrar so I also tried 7z: both versions won’t work so now I suppose that zip under Linux has different compression methods than either rar or 7z in windows.. I don’t know… maybe Linux zip uses lzma method while 7z winrar or winzip use some other…
I’ll give a try on my raspberry or another Linux pc I have…
I want to understand this error…
Yesterday I tested my script to wavpack Pekela organ by p. Grabowsky. It is incredible: total size of 1.5gb shrinks to 300mb. Now I have to understand how to correctly zip it. Maybe some parameter to be used otherwise I don’t understand why Pitea organ repacked doesn’t work. Il try with it under Linux…
For example, gzip usually builds archives with 32kb blocks: maybe GO is not able to read such method..

**Reply by jason-barba** [reply]
Apology accepted ;-) 
Have you tried to compare the .orgue file created by you with the original? This is a shot in the dark ...

**Reply by jason-barba** [reply]
Also, are you sure the organindex.ini is in the .orgue archive you created? Another shot ...

**Reply by paolo2504** [reply]
Yes and … yes !
This afternoon I’ll work on this ..😃

**Reply by jason-barba** [reply]
That's really weird than. Two identical files should be handled the same by the program.

**Reply by paolo2504** [reply]
Got it ! I'll keep you on thorns for a little bit more... It was really tricky thing... i'm ashamed..

**Reply by jason-barba** [reply]
You should be ashamed. I stroggled with GrandOrgueTool a while a go. @larspalo help me with the subtle difference between .\path\filename and path\filename. It was a lot of paths in the ODF I used ...

**Reply by oleg68** [reply]
> I have also tried winrar and 7zip (using 0-ratio compression that means store only) but GO is getting error on parsing content (“no end record found”).

Even the 0-ratio formats of winrar and 7zip formats differ the one used in `.orgue`

> As an alternative, I would greatly appreciate (in addition to have a working GrandOrgueTool) let GO read and manage directly the rar/zip files that we all find on internet.

It may be difficult with the license restrictions related to using rar format.

**Comment by larspalo** [comment]
@oleg68 Do you think we should add the information on how to build an .orgue package manually (https://github.com/GrandOrgue/grandorgue/discussions/955#discussioncomment-1939939) to the wiki page https://github.com/GrandOrgue/grandorgue/wiki/Organ-package-creation-(GrandOrgueTool) and split that page into two halves, the first with using GrandOrgueTool and the second with the manual instructions?

**Reply by oleg68** [reply]
I think, no, we should not. It is better to resolve #949

**Reply by paolo2504** [reply]
absolutely yes, even if i believe there is something strange in all stuff.
I have created a batch script to be run in Windows to wavpack recursively a sampleset. i've tested it and it works like a charm.
It's impressive the compression ratio wavpack achieves. For example, the sampleset Pekela got on P.Grabowsky site is reduced from 1.5gb to 367MB. The big problem is that using parameters you suggested (-h -x6) it took 2 hours to complete. I am using a smal laptop now with celeron n3060: for sure not very powerfull.. I'll test tomorrow with an i7 but i am sure that the process will be anycase slow.

Anyway, at present the big issue i have is that when i zip the file , both in Windows and linux (using the command you suggested with zip, GO is then not able to read parse it. It is not able to read .organ file even if it is there in the root of the sampleset.

I was able only to successfully zip your Pitea sampleset: i unpacked it the i repacked it again with 7z and now it works. No way with any other samplesets, even if i suspect the problem relies on the path of organindex.ini and the root structure.
It seems that organindex.ini and the .organ file must reside both on the root; additionally, no subfolders must exists. There is also a problem with the sub-path. For example, in Burea_funeral... all reference are like Pipe001=.\Subbas16\036-C.wav , not Pipe001=Subbas16\036-C.wav as you suggest.
Some hint required pls...

**Reply by larspalo** [reply]
> For example, in Burea_funeral... all reference are like Pipe001=.\Subbas16\036-C.wav , not Pipe001=Subbas16\036-C.wav as you suggest.

They all (the paths) need to be changed like I suggested. I of course noticed that while testing the errors @jason-barba got with GrandOrgueTool.

Pay **very** close attention to exactly how the .zip is created. The whole sampleset should be zipped in the root with no additional subfolder appearing first (ie. not having first BureaFuneralChapel as the first directory in the .zip file - the whole root directory of the sampleset **must** be directly available in the zip file and there the .organ file and organindex.ini file **must** be present (in the root). Next thing is the archive only - no compression extra with -0. I repeat the point nbr 4 above:

4. Pack the whole sampleset root directory into a zip archive. For instance one can (in the root directory) issue the following command:
`zip -r ../BureaFuneralChapel . -0`

I attach two pictures of how the .zip archive layout should be, and not be so that you can compare. I'm pretty sure your .zip layout is wrong.
![CorrectZipLayout](https://user-images.githubusercontent.com/6173941/148997107-3bf1c532-b827-4457-89a9-cbc014a785f0.png)
![IncorrectZipLayout](https://user-images.githubusercontent.com/6173941/148997153-83b4b302-7de8-4a19-bddd-fce296b55699.png)

**Reply by paolo2504** [reply]
Nono, my zip layout is correct: today, after much testing , I finally understood what you say. 
However, it’s not clear where and how the .organ file should be changed. Basically I don’t change it: I only build and add to the root the organ index file. 
Pls, which modifications have to be made to the .organ file ?

Also, you mentioned some testing by user Jason but, please, keep in mind I don’t have any documentation on grandorguetool, I cannot run it with option -h so I don’t really have many information. 
from post you so gently created for us in wiki section i now know that grandorguetool is able to wavpack files… again, I had to download wavpack r5.40 (latest) and build myself a batch script to do the job…
You speak of -o1 -o2 etc… parameters but I don’t know what they do…

finally, I am really interested in knowing how fast is packing with grandorguetool because the speed of wavpack, as far as I see, is too much slow….

**Reply by larspalo** [reply]
The .organ file should pass a "Perform strict ODF" check in the options without any log issues to begin with. (Any sampleset producer should always have that option enabled) Correct any reported log issues first. Then search and replace any `=.\` or `=./` paths to just `=`.

A very common issue for many windows users is case sensitivity (or rather the in-sensitivity)... Be sure to check that, too!

Unless I have access to the whole sampleset I cannot test it myself and say for sure exactly what needs to be changed, but the above notes "should" be enough...

**Reply by paolo2504** [reply]
I read right now your reply. Thanks. I was testing and i can say now it is FIXED. Problem was that i didn't realize that many if not almost every samplesets i used come with .organ files with path with "Linux/Unix" sysntax. For example Path1/Path2/Path3 while in Windows backslash is used that is  Path1\Path2\Path3. In addition many entries were preceeded by .\, thus making impossible or GO locate path.

aufff... Now i can use winrar to build packages. Tomorrow i'll test my script for wavpacking samplesets and try to set best parameters to speedup process that now is really too much slow.
If it took 2 hours to pack wav files with Pekela organ (1.5gb) i cannot even imagine how much it would take for example with Friesach (17.5gb).

**Reply by jason-barba** [reply]
Great.

**Comment by oleg68** [comment]
There are some feature requests here for the known organ list. It's a time to think how the features should be presented from the GUI point of view.

After merging the ``Organs`` and ``Organ Packages`` tabs the new tab looks like

![settings_organs tab](https://user-images.githubusercontent.com/11586506/149658907-9673f8d8-8c3e-4543-a781-6cf5ef9a4fb7.png)

Because an organ has lots of attributes and all attributes can not fit in a single line of a table, I'm going to do the classic master-detail view.

The upper part of the page will contain the table of organs (both packaged and unpackaged in the same table). Only few properties will be shown in the table: Name and a few other.

The downer part - various fields displaying all other properties of the organ selected in the upper part.

At the bottom of the page will be the buttons for the different actions on the selected organs.

Some actions (for example, `scan cache`) will not correspond the current organ.

My questions:

1. What columns should be in the table of organs?
2. What other properties should be shown in the downer part?
3. What actions should be available as buttons?

**Reply by larspalo** [reply]
1. I think that most of the info that can be found in the (loaded) organ properties dialog instead could be removed. Thus, we should keep: Church (name), MIDI, Organ package in the upper part, I think. It should perhaps be called: "Registered organs/packages"
2. For the lower information details view we could for instance have: Path, Id, Caches, Combinations/Settings. The info column doesn't seem to have anything there mostly at the moment. Thinking of it, I don't think the lower part should be implemented as a list at all, but as a panel with different widgets (maybe with separate listboxes for caches and settings files related to that organ)
3. The buttons for re-arranging the (upper) list order should be re-worked in the implementation so that the moved item will be visible in the list at all times and not disappearing like now. Ideally I'd like the possibility to delete an organ (together with everything related to it!) separately from the option of deleting any specifically selected cache or settings/combination file. The handling of deleting organs in a package vs not maybe needs some thought though (offer the option to remove the whole package or not). A fast and simple option for cleaning/removing all detected orphaned files/organs/packages would be a great addition too.

**Comment by paolo2504** [comment]
As suggested by Oleg I put here a new request. 
I would like to have Ctrl/Shift enabled to make multiple selection when deleting organs / packages in the Settings dialog. 
thanks.

**Comment by paolo2504** [comment]
I don't know if this is the right place to post enhancement request... anyway.. 

I would like to have ascending/descending sort capability on any column in the Settings/Organs tab.
regards.
paolo

**Comment by paolo2504** [comment]
question: are you eventually in mind to renew also the GO system bar itself ? In that case, i would personally appreciate having a couple of new icons on the left : Load Organs and Install Organ Package for example (just before 'Set' and Memory level.

![immagine](https://user-images.githubusercontent.com/50132661/153934469-d3ca97ad-7dd9-44b0-9579-425a7c2a1098.png)

thanks

**Reply by oleg68** [reply]
This topic only relates to the Organs tab. Please  open a separate topic for discussing enchancements the main toolbar

**Comment by ahall41** [comment]
- Make it easier to move organs up/down the list by dragging, rather than repeatedly hitting Up.
- If you have a Top button, there should also be one for Bottom. 
- Purge all MISSING organs
- The ability to edit the labels might be useful
- I use Open | Recent a lot. Please keep it!

**Comment by oleg68** [comment]
The work is in progress now. This is an intermediate variant of the Organs tab:

![image](https://user-images.githubusercontent.com/11586506/173222229-8eea37f0-d418-4509-ac46-a1fa6b558dbd.png)

**Comment by paolo2504** [comment]
Please, what organ hash, package hash, package id, package info stand for ?
tks

**Comment by oleg68** [comment]
Organ Hash corresponds the name of the cache and the preset files.
Package id is used for matching between Organs and Packages now.
Package hash seems is not used now.

**Comment by oleg68** [comment]
@paolo2504 

You have submitted most of issues about the Organs list.

Could you try the intermediate version of GrandOrgue with the Organs tab reworked https://github.com/oleg68/GrandOrgue-official/actions/runs/2602469146?

**Comment by paolo2504** [comment]
Ok.

**Reply by paolo2504** [reply]
Here below attached my thoughts in 12 points (at the beginning I copied last q&a between Oleg and myself).

[hints on new reworked setting in GrandOrgue.txt](https://github.com/GrandOrgue/grandorgue/files/9040699/hints.on.new.reworked.setting.in.GrandOrgue.txt)

**Comment by paolo2504** [comment]
Oleg, after further review i found a couple of issues on delete functions. Here follows detail.

a) Delete cache: cache files are deleted WITHOUT confirmation
b) Delete preset: here confirmation Yes/No is correctly shown but the word 'delete' is missing in the message
c) Delete: caches are deleted only AFTER clicking on Ok to exit the Settings dialog. Actually, cache files are correctly deleted. I just wonder if 
   the deletion after click on Ok button is by design (note that Delete cache and Delete preset are carried on immediately.
   Shortly speaking: 
   Delete cache and preset -> cache files are deleted and a log mesagge dialog is shown;
   Delete -> nothing happens until Ok button is selected: after that cache files are deleted and the log message is shown.
   Note: when Delete is selected, organs disappear from the list. If then you press Cancel, caches are not deleted and reentering Settings 
   organs are still there. This is correct.
d) Delete: .cache and .cmb files are deleted, .idx ones not. I think this has to be fixed.
    Note: .idx files are created only with package organs. I found this little bug only because i only use packages otherwise i would have not 
    noticed it.

**Reply by oleg68** [reply]
> Oleg, after further review i found a couple of issues on delete functions. Here follows detail.
> 
> a) Delete cache: cache files are deleted WITHOUT confirmation 

Because the cache files can be easy restored without a big problem

b) Delete preset: here confirmation Yes/No is correctly shown but the word 'delete' is missing in the message 

already fixed

c) Delete: caches are deleted only AFTER clicking on Ok to exit the Settings dialog. Actually, cache files are correctly deleted. I just wonder if the deletion after click on Ok button is by design (note that Delete cache and Delete preset are carried on immediately. Shortly speaking: Delete cache and preset -> cache files are deleted and a log mesagge dialog is shown; Delete -> nothing happens until Ok
 button is selected: after that cache files are deleted and the log message is shown. Note: when Delete is selected, organs disappear from the list. If then you press Cancel, caches are not deleted and reentering Settings organs are still there. This is correct.
 
It is by design: otherwise Cancel could not revere the side effect of Delete.
 
  d) Delete: .cache and .cmb files are deleted, .idx ones not. I think this has to be fixed. Note: .idx files are created only with package organs. I found this little bug only because i only use packages otherwise i would have not noticed it.

Yes, it is to be fixed.

**Comment by paolo2504** [comment]
I forgot mention that even Top, Down, Up functions are now ok. Good job.

**Comment by oleg68** [comment]
@paolo2504 @larspalo Thank you for your feedbacks.

We should decide now, what is to change before #1141 may be merged.

The main points relate to deletion of cache and presets:
- [x] The buttons are enabled regardless the actual existance of the files. May be to check? Just caching would not be sufficient: we also have to remember changes after deletion and relocations.
- [x] These buttons are the only that act immediately. In contrast all other actions that are performed only after pressing to `OK` and are not performed after Cancel. May be we should not delete cache and presets until `OK`?
- [x] Lack of deletion of orphaned caches and presets.

A possible solution is to discovering remember all existing preset and cache files on opening this dialog and marking what are to keep or not to keep. The result may be used for proper enabling/disabling the button. Theis task is not so simple: we need to take into account previous pressing to Delete and Relocate buttons. Pressing the delete cache/presets button only is to change the mark. The actual deletion will be only after exiting the dialog with `OK`. If several organs are selected, the delete cache/presets is to be enabled if ANY of the organ has existing cache/presets file.

@larspalo @paolo2504, is this idea acceptable? I would implement it. It will take about a week for me.

**Reply by larspalo** [reply]
I absolutely think this is acceptable.

**Comment by oleg68** [comment]
Another to do list:

- [ ] Add explanation of organ fields to the help
- [x] Separate the list of packages to delete in the confirmation question with new lines instead of commas
- [ ] Add the `Del` hotkey for deletion
- [x] Automatically focuse a line in the organ list and to show the properties after open and after delete
- [x] Delete index files after deleting packages

**Reply by larspalo** [reply]
Very good further improvements!

**Comment by oleg68** [comment]
@larspalo 
> Nor is the additional organs from a package removed when I remove the main organ of the package, but if I then remove the additional organ it asks me if I want to remove the main organ too - which I already had done. This could perhaps be improved upon.

There is a many-to-one relationship between organs and organ packages. There is no any `main` organ. After you delete all the organs from the package, you will be asked, if you want to remove the **package**, not the main organ. And you do not have an option not to delete the package: only not to delete organs that reference this package. Any orphaned packages are removed automatically.

**Reply by larspalo** [reply]
Right, it was poorly expressed by me - in that case, one of the organs (that I consider "main") had the same name as the package. However, it still is a possible confusion point we somehow need to clarify and document well.

**Comment by paolo2504** [comment]
Oleg, a couple of thoughts still on deletion.

1. Deleted Packages are not actually deleted.
Lars correctly says there is a many-to-one relationship between organs and organ packages. After you delete all the organs from the package, you will be asked, if you want to remove the package, not the main organ. This is absolutely correct, nothing to say on this.
Nevertheless, i would note that packages are authomatically reloaded/registered if you put them (i mean, the orgue files) in the folder that you defined as "organ packages" in the Paths tab.
This is causing a misleading behaviour imho; if you delete a package in Settings/Organs then you reopen Settings they are still there. The same with File/Load. Definitely, .cache, .cmb, presets, .idx (when fixed) are correctly deleted but the packages are still there.
Solution: eliminate the option to auto recognize package organs.
I know it is an early feature in GO but i don'ts see at present the advantage of maintaining it. Eliminating this feature you should also avoid the other issue regarding having many big .orgue files in the Package path: in this case, the first time you launch GO it could take hours to complete. This is the reason i always set an empty folder for packages and keep them in another location.

2. i personally woud enjoy having a new column "Package name" in the upper side of Settings/Organs tab. 
   This to allow see which organs belong to which package. At present the organ name can be see only in the fields below.
   It could be usefull for example when deleting organs to see how many organs still belong to which package.

**Reply by larspalo** [reply]
@paolo2504 @oleg68 I think that we definitely should keep the automatic registration of packages from the packages folder - it's by far the easiest way for new/casual user of GO to start playing with a certain .orgue package without any issues.

For more advanced users, such as yourself perhaps?, they should already know that if they simply remove the package from the packages folder and place them elsewhere and instead "install" the package - they can keep listing only what organs from a package they want.

I'm quite satisfied with seeing what organ belong to which package down below in the information textfields. The only issue I ran into is if the package was registered with an earlier version of GO they didn't contain that information, but it's easily fixed with just re-registering it.

All such issues described can easily be addressed with proper documentation of expected behavior and best practices.

**Reply by paolo2504** [reply]
Thanks for reply Lars. Regarding automatic registration i know you love it. Me too at the beginning.. I don't think it's a matter of being or not adv user... It is a fact that if you put tens of big packages in a folder and set that folder as Package path, then you get into a big trouble first time you launch GO. I tested this many times before understanding the reason, and every time i ended up in killing the running GO application from task manager. As an old developer, i don't like this and i would prefer to find a solution in the coding itself. Anyway you're right saying the solution is simple: just keep packages in a folder and NOT set it as Package folder (that's a nonsense by me).

Just a final note. With automatic registration package organs deleted with the new Delete button only disappear until the Settings tab is closed but still show up when Settings is reopened. A non-advanced user should not understand well this behaviour.
regards.

**Comment by oleg68** [comment]
@larspalo @paolo2504 I've finished working on the Organs tab. Could you try the new version from https://github.com/oleg68/GrandOrgue-official/actions/runs/2682478835 ?

> 1. Nevertheless, i would note that packages are authomatically reloaded/registered if you put them (i mean, the orgue files) in the folder that you defined as "organ packages" in the Paths tab.

Now `Delete` detects this case and brings the warning box.

**Reply by paolo2504** [reply]
Three notes.

1. DEL key doesn't work (i understood it was enabled)
2. .idx files are now deleted but only if .orgue files are outside the Package organ folder. For example, DEMO organ idx is delete (but it 
    doesn't reside in the packages folder...); also, package organs saved outside correctly get .idx files deleted.
3. Error 5: access denied is logged when deleting open package organs. In this case, both .cache and .idx files are not removed. 
   I suggest checking if any organ is open and if it is one of those going to be deleted: in this case not to proceed with deleting 
   that specific organ. 
   note: problem is that caches and idxs become orphan files because their matching organs don't show up anymore in 
   Settings.

note: .cache files of samplesets placed in the Sampleset folder are correctly deleted even if the organs are open.
Definitely, the access denied error arise only with package organs. We have two cases: 1) p.organs in package folder; 2) p.organs outside package folder. In both cases, .cache and .idx files are not removed if they are open during delete (error5 arises); what differs is that case 1) .idx and .cache files are rebuilt next time GO restarts (due to the auto discovery) while in case 2) the organs are removed from Settings and .idx and .cache become orphaned files.

**Reply by oleg68** [reply]
1. Confirm. It works on Linux and OsX, but not on Windows. I'll research.
2. The .idx file is deleted when you press OK file, but it appears again when GrandOrgue starts next time, because the package is autoregistered. It is desired behaviour.

> the organs are removed from Settings and .idx and .cache become orphaned files.

When you enter the Organ Settings and exit with `OK` the orphaned files will be deleted.

I'd not like to change anything with deleting an open packaged organ.

**Reply by paolo2504** [reply]
I know cache files are deleted after clicking Ok.
problem is that if you try to delete an organ that is open you get error 5 access denied if it’s a package (no error if it’s a sampleset).

When GO is restarted, idx is recreated but this happens only for packages that are located in the package folder. Only those packages are auto registered. 
Problem arises when package are outside the package folder. In this case, idx and cache are never recreated. As you have deleted them, they don’t appear anymore in Settings/Organs. Definitely the cache and idx files become orphan. 

You can reproduce this way:
1. Install a package that is located outside the package folder
2. Load it
3. Open Settings/organs and delete that organ
4. press Ok and you get the error5 access denied in the log window. 
5. Close Go and restart; then open Settings/organs. Correctly , the organ is not there because only packages in the defined folder are auto registered. 
6. Now check the GrandorgueCache folder and you’ll find the orphaned files. 

what I suggest Is simple: check if organs going to be deleted are open ; in that case, if they’re located in the package folder do nothing and proceed otherwise ask user to close the organ before deleting it (or force close before deletion).
![error acce denied](https://user-images.githubusercontent.com/50132661/179618888-87eb3848-9ca4-4c3e-b95f-d4ba0fcfce29.jpg)

**Comment by oleg68** [comment]
@paolo2504 I finished working on reorganising the Organs tab. The official intermediate version is 3.7.0-0.3 available for download at https://github.com/GrandOrgue/grandorgue/actions/runs/2762988899

The only problem that still persists is the error when deleting the organ that is currently open from the cache file. I created a separate issue for managing cache of the open organ on Windows: #1147. It requires a more complex solution.

The workaround of the cache orphaned files is to close the current organ or to restart GrandOrgan, to go to File->Settings and to press `OK`. The old orpaned cache files will be deleted.

**Reply by rousseldenis** [reply]
> @paolo2504 I finished working on reorganising the Organs tab. The official intermediate version is 3.7.0-0.3 available for download at https://github.com/GrandOrgue/grandorgue/actions/runs/2762988899
> 
> The only problem that still persists is the error when deleting the organ that is currently open from the cache file. I created a separate issue for managing cache of the open organ on Windows: #1147. It requires a more complex solution.
> 
@oleg68 Is there a mean to know which organ is open and to deactivate the organ selection for delete?

**Reply by oleg68** [reply]
Now it is not possible without a lot of coding: the Settings dialog does not know which organs are open.
